### PR TITLE
ARROW-9711: [Rust] Add new benchmark derived from TPC-H

### PR DESCRIPTION
**Pull request by andygrove**
This PR builds on https://github.com/apache/arrow/pull/7951.

Adds a new benchmark derived from TPC-H. I have been able to run this with a 100GB data set on my desktop and performance is about the same as Apache Spark running in local mode with the same thread count (24).

```
+--------------+--------------+-----------------+----------------------+--------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------+--------------------+----------------------+----------------------+-----------------+
| l_returnflag | l_linestatus | sum(l_quantity) | sum(l_extendedprice) | sum(l_extendedprice Multiply CAST(Int64(1) as Float64) Minus l_discount) | sum(l_extendedprice Multiply CAST(Int64(1) as Float64) Minus l_discount Multiply CAST(Int64(1) as Float64) Plus l_tax) | avg(l_quantity)    | avg(l_extendedprice) | avg(l_discount)      | count(UInt8(1)) |
+--------------+--------------+-----------------+----------------------+--------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------+--------------------+----------------------+----------------------+-----------------+
| A            | F            | 3775127758      | 5660776097194.464    | 5377736398183.935                                                        | 5592847429515.93                                                                                                       | 25.49937060623502  | 38236.11838745711    | 0.05000224145223291  | 148047881       |
| N            | F            | 98553062        | 147771098385.98004   | 140384965965.03473                                                       | 145999793032.77594                                                                                                     | 25.501475096542002 | 38237.03209968505    | 0.0499850931498342   | 3864590         |
| N            | O            | 7651423419      | 11473321691083.244   | 10899667121317.215                                                       | 11335664103186.313                                                                                                     | 25.499799813085986 | 38236.99077003657    | 0.04999757591275955  | 300058170       |
| R            | F            | 3775724970      | 5661603032745.35     | 5378513563915.415                                                        | 5593662252666.921                                                                                                      | 25.500067651651772 | 38236.70005754084    | 0.050001305269911714 | 148067261       |
+--------------+--------------+-----------------+----------------------+--------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------+--------------------+----------------------+----------------------+-----------------+
Query 1 iteration 2 took 21004 ms
```

### CODE REVIEW

**Comment by andygrove** [comment]
@jorgecarleitao @alamb fyi

**Comment by andygrove** [comment]
@wesm I believe that we can use TPC tests under their fair use policy [1] but we need to be careful to refer to them as "derived from" TPC since they are not official TPC tests.

[1] http://www.tpc.org/tpc_documents_current_versions/pdf/tpc_fair_use_quick_reference_v1.0.0.pdf

**Comment by andygrove** [comment]
@jorgecarleitao @alamb fyi

File: rust/benchmarks/README.md [hunk 1]
```diff
@@ -19,12 +19,27 @@
...
-Currently, only DataFusion benchmarks exist, but the plan is to add benchmarks for the arrow, flight, and parquet crates as well.
+Currently, only DataFusion benchmarks exist, but the plan is to add benchmarks for the arrow, flight, and parquet 
+crates as well. 
+
+## Benchmark derived from TPC-H
+
+These benchmarks are derived from the [TPC-H](http://www.tpc.org/tpch/) benchmark.
+
+```bash
+cargo run --release --bin tpch -- --iterations 3 --path /mnt/tpch/csv --format csv --query 1 --batch-size 4096
```
Comment (alamb) [inline rust/benchmarks/README.md:34]: I may have missed it, but if you could include instructions / links to instructions on how to actually create the TPCH data that would be super helpful

Comment (andygrove) [reply rust/benchmarks/README.md:35]: Done.

**Comment by andygrove** [comment]
@alamb @jorgecarleitao This is ready for review and is my last planned PR for the moment.

**Review by alamb — COMMENTED** [review]
LGTM
image

**Review by jorgecarleitao — COMMENTED** [review]
I believe that a bash script that would prepare the data for this would help to reproduce this benchmark. I would struggle to run it. I would not consider this a blocker at this point.

**Comment by andygrove** [comment]
@jorgecarleitao I can document it more .. it is actually pretty simple, but a good topic for a blog post perhaps.

**Pull request closed without merging**